// @prisma/language-server/dist/lib/completions/generator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatorSuggestions = void 0;
exports.getSuggestionForGeneratorField = getSuggestionForGeneratorField;
const vscode_languageserver_1 = require("vscode-languageserver");
const completions = __importStar(require("./completions.json"));
const ast_1 = require("../ast");
const internals_1 = require("./internals");
const klona_1 = require("klona");
const listAllAvailablePreviewFeatures_1 = __importDefault(require("../prisma-schema-wasm/listAllAvailablePreviewFeatures"));
/**
 * ```prisma
 * generator client {
 *  |
 * }
 * ```
 */
const supportedGeneratorFields = (0, internals_1.convertToCompletionItems)(completions.generatorFields, vscode_languageserver_1.CompletionItemKind.Field);
// generator.provider
const generatorProviders = (0, internals_1.convertToCompletionItems)(completions.generatorProviders, vscode_languageserver_1.CompletionItemKind.Constant);
/**
 * ```prisma
 * generator client {
 *   provider = "|"
 * }
 * ```
 */
const generatorProviderArguments = (0, internals_1.convertToCompletionItems)(completions.generatorProviderArguments, vscode_languageserver_1.CompletionItemKind.Property);
// generator.previewFeatures
const previewFeaturesArguments = (0, internals_1.convertToCompletionItems)(completions.previewFeaturesArguments, vscode_languageserver_1.CompletionItemKind.Property);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   runtime = |
 * }
 * ```
 */
const runtimeArguments = (0, internals_1.convertToCompletionItems)(completions.runtimeArguments, vscode_languageserver_1.CompletionItemKind.Property);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   runtime = "|"
 * }
 * ```
 */
const runtimes = (0, internals_1.convertToCompletionItems)(completions.runtimes, vscode_languageserver_1.CompletionItemKind.Constant);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   moduleFormats = |
 * }
 * ```
 */
const moduleFormatsArguments = (0, internals_1.convertToCompletionItems)(completions.moduleFormatArguments, vscode_languageserver_1.CompletionItemKind.Property);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   moduleFormats = "|""
 * }
 * ```
 */
const moduleFormats = (0, internals_1.convertToCompletionItems)(completions.moduleFormats, vscode_languageserver_1.CompletionItemKind.Constant);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   generatedFileExtension = |
 * }
 * ```
 */
const generatedFileExtensionArguments = (0, internals_1.convertToCompletionItems)(completions.generatedFileExtensionArguments, vscode_languageserver_1.CompletionItemKind.Property);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   generatedFileExtension = "|""
 * }
 * ```
 */
const generatedFileExtensions = (0, internals_1.convertToCompletionItems)(completions.generatedFileExtensions, vscode_languageserver_1.CompletionItemKind.Constant);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   importFileExtension = |
 * }
 * ```
 */
const importFileExtensionArguments = (0, internals_1.convertToCompletionItems)(completions.importFileExtensionArguments, vscode_languageserver_1.CompletionItemKind.Property);
/**
 * ```prisma
 * generator client {
 *   provider = "prisma-client"
 *   importFileExtension = "|""
 * }
 * ```
 */
const importFileExtensions = (0, internals_1.convertToCompletionItems)(completions.importFileExtension, vscode_languageserver_1.CompletionItemKind.Constant);
// Fields for prisma-client generator
const prismaClientFields = (0, internals_1.convertToCompletionItems)(completions.generatorPrismaClientFields, vscode_languageserver_1.CompletionItemKind.Field);
// Fields for prisma-client-js generator
const prismaClientJSFields = (0, internals_1.convertToCompletionItems)(completions.generatorPrismaClientJSFields, vscode_languageserver_1.CompletionItemKind.Field);
/**
 * Helper function to get the provider value from a generator block
 */
function getProviderFromBlock(block) {
    const lines = block.definingDocument.lines;
    for (let lineIndex = block.range.start.line + 1; lineIndex < block.range.end.line; lineIndex++) {
        const line = lines[lineIndex].text.trim();
        if (line.startsWith('provider')) {
            // given a line like `provider = "prisma-client"`, extract the value of the provider (e.g., 'prisma-client').
            const match = line.match(/\s*provider\s*=\s*"([^"]+)"/);
            return match ? match[1] : undefined;
        }
    }
    return undefined;
}
function handlePreviewFeatures(previewFeaturesArray, position, currentLineUntrimmed, isInsideQuotation) {
    const previewFeatures = previewFeaturesArray.map((pf) => vscode_languageserver_1.CompletionItem.create(pf));
    if ((0, ast_1.isInsideAttribute)(currentLineUntrimmed, position, '[]')) {
        if (isInsideQuotation) {
            const usedValues = (0, ast_1.getValuesInsideSquareBrackets)(currentLineUntrimmed);
            return {
                items: previewFeatures.filter((t) => !usedValues.includes(t.label)),
                isIncomplete: true,
            };
        }
        return {
            items: previewFeaturesArguments.filter((arg) => !arg.label.includes('[')),
            isIncomplete: true,
        };
    }
    else {
        if (!isInsideQuotation) {
            return {
                items: previewFeaturesArguments.filter((arg) => !arg.label.includes('"')),
                isIncomplete: true,
            };
        }
        const firstQuoteIndex = currentLineUntrimmed.indexOf('"');
        const lastQuoteIndex = currentLineUntrimmed.lastIndexOf('"');
        const convertToListAndSuggestMore = {
            additionalTextEdits: [vscode_languageserver_1.TextEdit.insert(vscode_languageserver_1.Position.create(position.line, lastQuoteIndex + 1), ']')],
            command: {
                command: 'editor.action.triggerSuggest',
                title: 'Trigger Suggest',
            },
            kind: vscode_languageserver_1.CompletionItemKind.Value,
            label: '[""] (convert to list)',
            insertTextFormat: vscode_languageserver_1.InsertTextFormat.Snippet,
            insertTextMode: vscode_languageserver_1.InsertTextMode.asIs,
            preselect: true,
            textEdit: vscode_languageserver_1.TextEdit.insert(vscode_languageserver_1.Position.create(position.line, firstQuoteIndex - 1), ' ["'),
        };
        return {
            items: [convertToListAndSuggestMore],
            isIncomplete: true,
        };
    }
}
/**
 * Removes all field suggestion that are invalid in this context. E.g. fields that are used already in a block will not be suggested again.
 * This function removes all field suggestion that are invalid in a certain context. E.g. in a generator block `provider, output, platforms, pinnedPlatForm`
 * are possible fields. But those fields are only valid suggestions if they haven't been used in this block yet. So in case `provider` has already been used, only
 * `output, platforms, pinnedPlatform` will be suggested.
 */
function removeInvalidFieldSuggestions(supportedFields, block, schema, position) {
    let reachedStartLine = false;
    for (const { lineIndex, text } of schema.iterLines()) {
        if (lineIndex === block.range.start.line + 1) {
            reachedStartLine = true;
        }
        if (!reachedStartLine || lineIndex === position.line) {
            continue;
        }
        if (lineIndex === block.range.end.line) {
            break;
        }
        const fieldName = text.replace(/ .*/, '');
        if (supportedFields.includes(fieldName)) {
            supportedFields = supportedFields.filter((field) => field !== fieldName);
        }
    }
    return supportedFields;
}
function getSuggestionForGeneratorField(block, schema, position) {
    const provider = getProviderFromBlock(block);
    let suggestions;
    if (provider === 'prisma-client-js') {
        suggestions = (0, klona_1.klona)(prismaClientJSFields);
    }
    else if (provider === 'prisma-client') {
        suggestions = (0, klona_1.klona)(prismaClientFields);
    }
    else {
        // Default to showing all fields if no provider is set yet
        suggestions = (0, klona_1.klona)(supportedGeneratorFields);
    }
    const labels = removeInvalidFieldSuggestions(suggestions.map((item) => item.label), block, schema, position);
    return suggestions.filter((item) => labels.includes(item.label));
}
const generatorSuggestions = (line, untrimmedLine, position, isInsideQuotation, onError) => {
    /**
     * Common cases for generator blocks:
     */
    // provider
    if (line.startsWith('provider')) {
        const providers = generatorProviders;
        if (isInsideQuotation) {
            return {
                items: providers,
                isIncomplete: true,
            };
        }
        return {
            items: generatorProviderArguments,
            isIncomplete: true,
        };
    }
    // previewFeatures
    if (line.startsWith('previewFeatures')) {
        const generatorPreviewFeatures = (0, listAllAvailablePreviewFeatures_1.default)(onError);
        if (generatorPreviewFeatures.length > 0) {
            return handlePreviewFeatures(generatorPreviewFeatures, position, untrimmedLine, isInsideQuotation);
        }
        return undefined;
    }
    /**
     * Cases for `provider = "prisma-client"`
     */
    // runtime
    if (line.startsWith('runtime')) {
        if (isInsideQuotation) {
            return {
                items: runtimes,
                isIncomplete: true,
            };
        }
        return {
            items: runtimeArguments,
            isIncomplete: true,
        };
    }
    // runtime
    if (line.startsWith('moduleFormat')) {
        if (isInsideQuotation) {
            return {
                items: moduleFormats,
                isIncomplete: true,
            };
        }
        return {
            items: moduleFormatsArguments,
            isIncomplete: true,
        };
    }
    // generatedFileExtension
    if (line.startsWith('generatedFileExtension')) {
        if (isInsideQuotation) {
            return {
                items: generatedFileExtensions,
                isIncomplete: true,
            };
        }
        return {
            items: generatedFileExtensionArguments,
            isIncomplete: true,
        };
    }
    // importFileExtension
    if (line.startsWith('importFileExtension')) {
        if (isInsideQuotation) {
            return {
                items: importFileExtensions,
                isIncomplete: true,
            };
        }
        return {
            items: importFileExtensionArguments,
            isIncomplete: true,
        };
    }
};
exports.generatorSuggestions = generatorSuggestions;
